native/common/include/jp_primitive_accessor.h

```diff
@@ -164,16 +164,25 @@ class JPConversionLong : public JPConversion
 	virtual jvalue convert(JPMatch &match) override
 	{
 		jvalue res;
-		jlong val = PyLong_AsLongLong(match.object);
-		if (val == -1)
-			JP_PY_CHECK();
-		base_t::field(res) = (typename base_t::type_t) base_t::assertRange(val);
+		if (match.type == JPMatch::_exact)
+		{
+			jlong val = (jlong) PyLong_AsUnsignedLongLongMask(match.object);
+			if (val == -1)
+				JP_PY_CHECK();
+			base_t::field(res) = (typename base_t::type_t) val;
+		} else
+		{
+			jlong val = (jlong) PyLong_AsLongLong(match.object);
+			if (val == -1)
+				JP_PY_CHECK();
+			base_t::field(res) = (typename base_t::type_t) base_t::assertRange(val);
+		}
```

Comment:
We select the type of conversion here.  We need different functions based on the whether it is an implicit or a user requested cast.